content/blog/2022-03-whats-new-in-axum-0-5.md

```diff
@@ -0,0 +1,298 @@
+from the request. But what you might not realise is that this would implicitly
+consume the headers such that other extractors wouldn't be able to access them.
+
+For example this is subtly broken:
+
+```rust
+use axum::{http::HeaderMap, extract::Form};
+
+async fn handler(
+    headers: HeaderMap,
+    form: Form<Payload>,
+) {
+    // ...
+}
+```
+
+Since we run the `HeaderMap` first, `Form` would be unable to access them and
+fail with a `500 Internal Server Error`. This was quite surprising and caused
+headaches for some users.
+
+However in axum 0.5 this problem goes away and it just works!
+
+## More flexible `Router::merge`
+
+[`Router::merge`] can be used to merge two routers into one. In axum 0.5 its
+gotten slightly more flexible since it now accepts any `impl Into<Router>`. This
+allows you to have custom ways of constructing `Router`s and have them work
+seamlessly with axum.
+
+One could imagine a way to compose REST and gRPC like so:
+
+```rust
+let rest_routes = Router::new().route(...);
+
+// with `impl From<GrpcService> for Router`
+let grpc_service = GrpcService::new(GrpcServiceImpl::new());
+
+let app = Router::new()
+    .merge(rest_routes)
+    .merge(grpc_service);
+```
+
+## Honorable mentions
+
+The following features weren't new in 0.5 but shipped recently and are worthy of
+a shout out.
+
+### `middleware::from_fn`
+
+axum uses the [`tower::Service`] trait for middleware however it can be little
+daunting to implement, mainly due to the lack of async traits in Rust.
+
+But with [`axum::middleware::from_fn`] you can hide all that complexity and use a
+familiar async function:
+
+```rust
+use axum::{
+    Router,
+    http::{Request, StatusCode},
+    routing::get,
+    response::IntoResponse,
+    middleware::{self, Next},
+};
+
+async fn my_middleware<B>(
+    req: Request<B>,
+    next: Next<B>,
+) -> impl IntoResponse {
+    // transform the request...
+
+    let response = next.run(request).await;
+
+    // transform the response...
+
+    response
+}
+
+let app = Router::new()
+    .route("/", get(|| async { /* ... */ }))
+    // add our middleware function
+    .layer(middleware::from_fn(my_middleware));
+```
+
+The [documentation for middleware] has also been reworked and goes into more
+details about the different ways to write middleware, when to pick which
+approach, how ordering works, and more.
+
+### Type safe routing
+
+In axum-extra we're experimenting with "type safe routing". The idea is is to
+establish a type safe connection between a path and the corresponding handler.
+
+Previously it was possible to add a path like `/users` and apply a `Path<u32>`
+extractor, which would always fail at runtime since the path doesn't contain any
+parameters.
+
+We can use axum-extra's type safe routing to make that problem go away:
+
+```rust
+use serde::Deserialize;
+use axum::Router;
+use axum_extra::routing::{
+    TypedPath,
+    RouterExt, // for `Router::typed_get`
+};
+
+// A type safe path
```

Comment:
in general, i think "type-safe" is normally hyphenated? but i could be convinced otherwise
```suggestion
### Type-safe routing

In axum-extra we're experimenting with "type-safe routing". The idea is is to
establish a type-safe connection between a path and the corresponding handler.

Previously it was possible to add a path like `/users` and apply a `Path<u32>`
extractor, which would always fail at runtime since the path doesn't contain any
parameters.

We can use axum-extra's type-safe routing to make that problem go away:

```rust
use serde::Deserialize;
use axum::Router;
use axum_extra::routing::{
    TypedPath,
    RouterExt, // for `Router::typed_get`
};

// A type-safe path
```